node/internal/fs/streams.ts

```diff
@@ -66,121 +144,231 @@ export function WriteStream(
 
   const self = this as WriteStream;
 
-  Writable.call(self, {
-    highWaterMark: options?.highWaterMark,
-    decodeStrings: options?.decodeStrings,
-    defaultEncoding: options?.defaultEncoding,
-    objectMode: options?.objectMode,
-    emitClose: options?.emitClose,
-    autoDestroy: options?.autoClose ?? options?.autoDestroy,
-    signal: options?.signal,
-  });
+  options = copyObject(getOptions(options, kEmptyObject));
+
+  // Only buffers are supported.
+  options.decodeStrings = true;
+
+  if (options.fd == null) {
+    self.fd = null;
+    self[kFs] = options.fs ||
+      { open: fsOpen, write: fsWrite, writev: fsWritev, close: fsClose };
+    validateFunction(self[kFs].open, "options.fs.open");
+
+    // Path will be ignored when fd is specified, so it can be falsy
+    self.path = toPathIfFileURL(path);
+    self.flags = options.flags === undefined ? "w" : options.flags;
+    self.mode = options.mode === undefined ? 0o666 : options.mode;
+
+    validatePath(self.path);
+  } else {
+    self.fd = getValidatedFd(importFd(self, options));
+  }
+
+  options.autoDestroy = options.autoClose === undefined
+    ? true
+    : options.autoClose;
+
+  if (!self[kFs].write && !self[kFs].writev) {
+    throw new ERR_INVALID_ARG_TYPE(
+      "options.fs.write",
+      "function",
+      self[kFs].write,
+    );
+  }
 
-  self.fd = null;
-  self.path = toPathIfFileURL(path);
-  self.flags = options?.flags ?? "w";
-  self.mode = options?.mode ?? 0o666;
+  if (self[kFs].write) {
+    validateFunction(self[kFs].write, "options.fs.write");
+  }
+
+  if (self[kFs].writev) {
+    validateFunction(self[kFs].writev, "options.fs.writev");
+  }
+
+  if (options.autoDestroy) {
+    validateFunction(self[kFs].close, "options.fs.close");
+  }
+
+  // It's enough to override either, in which case only one will be used.
+  if (!self[kFs].write) {
+    // TODO(PolarETech): Reduce @ts-ignore as much as possible.
+    // deno-lint-ignore ban-ts-comment
+    // @ts-ignore
+    self._write = null;
+  }
+  if (!self[kFs].writev) {
+    // TODO(PolarETech): Reduce @ts-ignore as much as possible.
+    // deno-lint-ignore ban-ts-comment
+    // @ts-ignore
+    self._writev = null;
+  }
```

Comment:
It is difficult to properly resolve type errors, so @ts-ignore is given.

In this part, assign `null` to not override `this._write()` or `this._writev()` in the parent class if `fs.write` or `fs.writev` is not passed as the `fs` option.
Changing the WriteStream interface to allow null for these functions causes a type error because it conflicts with the type declaration of the inherited Writable class.
It is difficult for me to solve this while keeping the implementation close to Node.